man/shmcat.man
.TH SHMCAT "1" "August 2016" "shmcat(1)"
.SH NAME
shmcat \- dump shared memory segments, files, standard input and/or text
.SH SYNOPSIS
.B shmcat [
.I OPTION
.B ]... [
.I OPERAND
.B ]...
.SH DESCRIPTION
Dump shared memory segments, files, standard input and/or text to standard
output. If no OPERAND is given, read standard input only. If more
than one OPERAND is given, dump them in order. If some of the OPERANDs
cannot be read, shmcat prints an error message to standard error and tries
to continue with the other OPERANDs, unless the \fB-S\fR or \fB--stop\fR
option is given.
.PP
Most modern systems support two APIs for shared memory:
.PP
The older one is commonly known as the System V API. It uses numeric keys
and IDs as identifiers and consists of system calls like \fBshmget(2)\fR,
\fBshmat(2)\fR and \fBshmdt(2)\fR.
.PP
The newer API is called the POSIX shared memory API. It uses names in string
form as identifiers. It treats shared memory similar to files, which are
opened with \fBshm_open(3)\fR and then mapped into the program address space
via \fBmmap(2)\fR. Please note that POSIX shared memory is not supported by all
versions of shmcat.
.PP
If shmcat has been compiled without \fBgetopt_long(3)\fR support, only the
short, one character OPTIONs and OPERANDs can be used. Otherwise the long ones
are also available. When a long OPTION or OPERAND needs an argument, the short
versions do so, as well.
.SH OPTIONS
.TP
\fB-h\fR, \fB-?\fR, \fB--help\fR
Display usage information and quit without dumping anything.
.TP
\fB-S\fR, \fB--stop\fR
By default, if an OPERAND cannot be read, shmcat tries to continue with the
remaining OPERANDs. This option turns off that behavior, so that the program
stops immediately after the first OPERAND that cannot be read.
.TP
\fB-V\fR, \fB--version\fR
Show version information and quit without further action.
.SH OPERANDS
.TP
\fB-f\fR, \fB--file=\fIFILENAME\fR
Dump the contents of the given file.
.TP
\fB-i\fR, \fB--stdin\fR
Dump standard input.
.TP
\fB-M\fR, \fB--shmkey=\fIKEY\fR
Dump the System V shared memory segment with the given key. The key can be
given as a decimal (no prefix), octal (prefix \fB0\fR), or hexadecimal
(prefix \fB0x\fR) number. It can, for example, be obtained using the
\fBipcs(1)\fR or the \fBftok(1)\fR tool.
.TP
\fB-m\fR, \fB--shmid=\fIID\fR
Dump the System V shared memory segment with the given id. The id can be
given as a decimal (no prefix), octal (prefix \fB0\fR), or hexadecimal
(prefix \fB0x\fR) number. It can, for example, be obtained using the
\fBipcs(1)\fR tool.
.TP
\fB-n\fR, \fB--newline\fR
Add a line feed.
.TP
\fB-p\fR, \fB--posix-shm=\fINAME\fR
Dump the POSIX shared memory segment with the given name (not supported in
all installations).
.TP
\fB-t\fR, \fB--text=\fITEXT\fR
Print the given text.
.SH EXIT CODES
.TP
0
No errors, every OPERAND requested was dumped.
.TP
1
There was an error while opening or reading at least one OPERAND. If shmcat
has been invoked with the \fB-S\fR or \fB--stop\fR option, the program stopped
after this error, otherwise it tried to continue with the remaining OPERANDs
(if given).
.TP
2
There was a fatal error, usually this means that shmcat could not write to standard output.
.TP
10
Wrong usage of the program.
.SH NOTES
Usually, when you dump shared memory segments, you should redirect the
output of shmcat to a file or a program like \fBod(1)\fR, because shared
memory data usually has a binary form and thus it may contain
non-printable characters. You also probably want to suspend the processes
writing to the shared memory object in order to get a consistent dump.
.SH AUTHOR
Stefan Gast
.SH "SEE ALSO"
.BR ftok(1),
.BR ipcs(1),
.BR shmat(2),
.BR shmget(2),
.BR shm_open(3)


NEWS
This is shmcat 1.9

shmcat-1.9 (2017-12-16)
	made -s,--stop work if immediately following operand fails
	switched from tar.bz2 to tar.xz release tarballs

shmcat-1.8 (2016-08-12)
	added a command line option to stop on first failed operand

shmcat-1.7 (2014-08-04)
	fixed command line parsing with getopt_long and abbreviated long options
	made building from git work under FreeBSD 10 (older versions still fail)

shmcat-1.6 (2012-11-26)
	fixed ftok printing the wrong keys with IDs > 127
	fixed shmcat not liking some keys
	POSIX shared memory support
	improved output of shmcat compile options
	fixed error messages in some rarely seen cases
	documentation fixes

shmcat-1.5 (2012-10-14)
	initial, official FreeBSD support
	ftok checks the command line more strictly
	documentation and build system fixes

shmcat-1.4 (2012-03-30)
	added the ftok utility to calculate IPC keys from the command line
	translated man pages into German
	several documentation fixes

shmcat-1.3 (2012-03-04)
	further increased buffer size
	added configure option to use smaller buffer
	added English man page

shmcat-1.2 (2011-07-14)
	increased buffer size for dumping files (better performance)
	made some debug messages more verbose
	support for systems with very large key_t data type

shmcat-1.1 (2011-02-11)
	fixed typos in some rarely seen error messages

shmcat-1.0 (2010-12-01)
	added German translation
	made shared memory key parsing more robust

shmcat-0.8 (2010-11-26)
	initial beta release


src/shmcat.c
/*
 * shmcat: Dump Shared Memory Segments and more
 * (C) 2012, 2014, 2016, 2017 by Stefan Gast
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * shmcat.c: shmcat main program
 */

#include "config.h"
#include <stdio.h>
#include <locale.h>
#include <unistd.h>
#include "translate.h"

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#include "status.h"
#include "filedumper.h"
#include "printtext.h"
#include "shmdumper.h"


enum
{
	EXITCODE_OK		= 0,
	EXITCODE_ERROR		= 1,
	EXITCODE_FATAL		= 2,
	EXITCODE_USAGE		= 10
};


static void usage(const char *progname)
{
	printf(_("Usage: %s [OPTION]... [OPERAND]...\n"), progname);
	puts(_("Print files, shared memory segments and/or the given text\n"
	       "to standard output.\n"));

	puts(_("Following OPTIONs are supported:\n"));

#ifdef HAVE_GETOPT_LONG
	puts(_("-h, --help               display this help and exit.\n"
	       "-?                       same as -h.\n"
	       "-S, --stop               stop immediately if there is an error with an operand.\n"
	       "-V, --version            output version information and exit.\n"));
#else /* !HAVE_GETOPT_LONG */
	puts(_("-h             display this help and exit.\n"
	       "-?             same as -h.\n"
	       "-S             stop immediately if there is an error with an operand.\n"
	       "-V             output version information and exit.\n"));
#endif /* HAVE_GETOPT_LONG */

	puts(_("An OPERAND may be one of:\n"));

#ifdef HAVE_GETOPT_LONG
	puts(_("-f, --file=filename      dump contents of the given file.\n"
	       "-i, --stdin              dump standard input.\n"
	       "-M, --shmkey=key         dump the System V shared memory segment with the\n"
	       "                         given key.\n"
	       "-m, --shmid=id           dump the System V shared memory segment with the\n"
	       "                         given id.\n"
	       "-n, --newline            add a line feed."));
#ifdef ENABLE_POSIX_SHM
	puts(_("-p, --posix-shm=name     dump the POSIX shared memory segment with the\n"
	       "                         given name."));
#endif /* ENABLE_POSIX_SHM */
	puts(_("-t, --text=text          print the given text.\n"));

#else /* !HAVE_GETOPT_LONG */
	puts(_("-f filename    dump contents of the given file.\n"
	       "-i             dump standard input.\n"
	       "-M key         dump the System V shared memory segment with the given key.\n"
	       "-m id          dump the System V shared memory segment with the given id.\n"
	       "-n             add a line feed."));
#ifdef ENABLE_POSIX_SHM
	puts(_("-p name        dump the POSIX shared memory segment with the given name."));
#endif /* ENABLE_POSIX_SHM */
	puts(_("-t text        print the given text.\n"));
#endif /* HAVE_GETOPT_LONG */

	puts(_("If no operand is given, read standard input."));
}


static void show_version(void)
{
	printf("%s (", PACKAGE_STRING);
#ifdef ENABLE_BIG_BUFFER
	printf("big-buffer");
#else /* !ENABLE_BIG_BUFFER */
	printf("small-buffer");
#endif /* ENABLE_BIG_BUFFER */
#ifdef ENABLE_POSIX_SHM
	printf(", posix-shm");
#endif /* ENABLE_POSIX_SHM */
	puts(")");

	puts(_("Copyright (C) 2012, 2014, 2016, 2017 by Stefan Gast"));
	puts(_("This is free software.  You may redistribute copies of it under the terms of\n"
               "the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.\n"
               "There is NO WARRANTY, to the extent permitted by law.\n"));

	/* TRANSLATORS: The placeholder indicates the bug-reporting address
           for this package.  Please add _another line_ saying
           "Report translation bugs to <...>\n" with the address for translation
           bugs (typically your translation team's web or email address).  */
	printf(_("Report bugs to: %s\n"), PACKAGE_BUGREPORT);
}


int main(int argc, char **argv)
{
#ifdef ENABLE_POSIX_SHM
	const char *optstr = "hSVf:iM:m:np:t:";
#else
	const char *optstr = "hSVf:iM:m:nt:";
#endif
#if defined HAVE_GETOPT_LONG && defined HAVE_GETOPT_H
	static const struct option long_options[] =
	{
		{ "help", no_argument, NULL, 'h' },
		{ "stop", no_argument, NULL, 'S' },
		{ "version", no_argument, NULL, 'V' },
		{ "file", required_argument, NULL, 'f' },
		{ "stdin", no_argument, NULL, 'i' },
		{ "shmkey", required_argument, NULL, 'M' },
		{ "shmid", required_argument, NULL, 'm' },
		{ "newline", no_argument, NULL, 'n' },
#ifdef ENABLE_POSIX_SHM
		{ "posix-shm", required_argument, NULL, 'p' },
#endif
		{ "text", required_argument, NULL, 't' },
		{ NULL, 0, NULL, 0 }
	};
#endif
	int opt;
	int stop = 0;
	ShmcatStatus status = SHMCAT_NOTHING_DONE;
	ShmcatStatus dumperStatus;

	/* Initialize i18n support */
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	/* Do a first run of getopt/getopt_long to check the command line for
	 * validity and to handle help or version info options. */
	opterr = 0;
	optind = 1;
#if defined HAVE_GETOPT_LONG && defined HAVE_GETOPT_H
	while((opt = getopt_long(argc, argv, optstr, long_options, NULL)) != -1)
#else
	while((opt = getopt(argc, argv, optstr)) != -1)
#endif
	{
		switch(opt)
		{
			case 'h':
				usage(argv[0]);
				return EXITCODE_OK;

			case 'S':
				stop = 1;
				break;

			case 'V':
				show_version();
				return EXITCODE_OK;

			case '?':
				if(optopt == '?')
				{
					usage(argv[0]);
					return EXITCODE_OK;
				}
				else
				{
					fprintf(stderr, _("%s: Wrong usage, try \"%s -h\" for more information!\n"), argv[0], argv[0]);
					return EXITCODE_USAGE;
				}
		}
	}
	if(argc != optind)
	{
		fprintf(stderr, _("%s: Wrong usage, try \"%s -h\" for more information!\n"), argv[0], argv[0]);
		return EXITCODE_USAGE;
	}

	/* Second getopt run does the actual work */
	optind = 1;
#if defined HAVE_GETOPT_LONG && defined HAVE_GETOPT_H
	while((opt = getopt_long(argc, argv, optstr, long_options, NULL)) != -1 &&
			status != SHMCAT_PANIC && !(stop && status == SHMCAT_ERROR))
#else
	while((opt = getopt(argc, argv, optstr)) != -1 &&
			status != SHMCAT_PANIC && !(stop && status == SHMCAT_ERROR))
#endif
	{
		switch(opt)
		{
			case 'f':
				dumperStatus = dumpFilename(optarg, argv[0]);
				break;
			case 'i':
				dumperStatus = dumpFd(STDIN_FILENO, NULL, argv[0]);
				break;
			case 'M':
				dumperStatus = dumpShmKey(optarg, argv[0]);
				break;
			case 'm':
				dumperStatus = dumpShmId(optarg, argv[0]);
				break;
			case 'n':
				dumperStatus = printText("\n", argv[0]);
				break;
#ifdef ENABLE_POSIX_SHM
			case 'p':
				dumperStatus = dumpPosixShm(optarg, argv[0]);
				break;
#endif
			case 't':
				dumperStatus = printText(optarg, argv[0]);
				break;
			case 'S':
				dumperStatus = SHMCAT_NOTHING_DONE;
				break;
			default:
				/* Notify the user someting is wrong, but do not handle this
				 * as a fatal error. */
				fprintf(stderr, _("%s: BUG: Unhandled command line option: %c\n"), argv[0], opt);
				dumperStatus = SHMCAT_ERROR;
				break;
		}

		/* The ShmcatStatus enum is ordered ascendingly, so we just have to
		 * increase our status variable if the current run of the loop led
		 * to a higher status value. */
		if(status < dumperStatus)
			status = dumperStatus;
	}

	/* If we haven't done anything until here, just dump stdin. */
	if(status == SHMCAT_NOTHING_DONE)
		status = dumpFd(STDIN_FILENO, NULL, argv[0]);

	switch(status)
	{
		case SHMCAT_OK:
			return EXITCODE_OK;
		case SHMCAT_ERROR:
			return EXITCODE_ERROR;
		case SHMCAT_PANIC:
			return EXITCODE_FATAL;
		default:
			fprintf(stderr, _("%s: BUG: One of the program routines returned an invalid value (%i)!\n"), argv[0], (int)status);
			return EXITCODE_ERROR;
	}
}
